### PR TITLE
Prometheus exporter: concurrent collect bugfix

### PR DESCRIPTION
**Pull request by hexdigest**
This PR fixes the following problem which is caused by the concurrent Collect method execution:

```
fatal error: concurrent map writes

goroutine 112 [running]:
go.opentelemetry.io/otel/exporters/prometheus.validateMetrics({0xc0004fc1e0, 0x28}, {0x0, 0x0}, 0xc00024e4c8, 0xc000122008?)
        /go/pkg/mod/go.opentelemetry.io/otel/exporters/prometheus@v0.36.0/exporter.go:384 +0x5fd
go.opentelemetry.io/otel/exporters/prometheus.addSumMetric[...](0x7f27d20dd108?, {{0xc0003693b0?, 0x40f7c7?, 0x28?}, 0x80?, 0xb8?}, {{0x1390626, 0x22}, {0x0, 0x0}, ...}, ...)
        /go/pkg/mod/go.opentelemetry.io/otel/exporters/prometheus@v0.36.0/exporter.go:214 +0xcd
go.opentelemetry.io/otel/exporters/prometheus.(*collector).Collect(0xc00009e730, 0xc0004f5f60?)
        /go/pkg/mod/go.opentelemetry.io/otel/exporters/prometheus@v0.36.0/exporter.go:160 +0xc6e
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
```

Also fixes https://github.com/open-telemetry/opentelemetry-go/issues/3917

### CODE REVIEW

File: exporters/prometheus/exporter.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ import (
 
 	"github.com/prometheus/client_golang/prometheus"
 	dto "github.com/prometheus/client_model/go"
+	"github.com/puzpuzpuz/xsync"
```
Comment (MrAlias) [inline exporters/prometheus/exporter.go:29]: I would rather not introduce a dependency just to deal with concurrency of a type.

**Review by MrAlias — DISMISSED** [review]
This is missing tests to validate the bug this is supposed to address is resolved.

File: exporters/prometheus/exporter.go (hunk 1 above)
Comment (Aneurysm9) [reply exporters/prometheus/exporter.go:29]: Doubly so when [`sync.Map` exists](https://pkg.go.dev/sync#Map).

**Comment by Aneurysm9** [comment]
Is this type expected to be safe for concurrent use?  I don't believe we document that any exporters are safe for concurrent use.  What is the scenario that causes this error to be observed?

File: exporters/prometheus/exporter.go (hunk 1 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go:29]: sync.Map isn't a type safe thing unlike this implementation based on generics. However, I agree that we should avoid new dependencies here. Do you want me to:
* use sync.Map
* protect the existing map with sync.RWMutex
* implement a generic map type in this library (potentially could be re-used in other places in this repo)

**Comment by hexdigest** [comment]
> Is this type expected to be safe for concurrent use? I don't believe we document that any exporters are safe for concurrent use. 

Neither prometheus nor opentelemetry explicitly documented anything about concurrency:

https://github.com/prometheus/client_golang/blob/27f0506d6ebbb117b6b697d0552ee5be2502c5f2/prometheus/registry.go#L140:L160

> What is the scenario that causes this error to be observed?

We use OpenTelemetry with Prometheus in an absolutely standard way. I think that it just happen that collector came concurrently which is very unlikely but still possible.

File: exporters/prometheus/exporter.go (hunk 1 above)
Comment (MrAlias) [reply exporters/prometheus/exporter.go:29]: Given the type looks to be passed I would expect a new concurrent map type is needed. That said, I don't know if it needs to be generic given it is for a single use.

File: exporters/prometheus/exporter.go [hunk 2]
```diff
@@ -122,16 +122,18 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		}
 	}
 
-	c.createTargetInfoOnce.Do(func() {
-		// Resource should be immutable, we don't need to compute again
-		targetInfo, err := c.createInfoMetric(targetInfoMetricName, targetInfoDescription, metrics.Resource)
+	c.mu.Lock()
```
Comment (hexdigest) [inline exporters/prometheus/exporter.go:125]: May be it will be easier to use lock for the whole Collect method not just critical sections?

**Comment by hexdigest** [comment]
@mrAlias @Aneurysm9 

Digging deeper I found out that concurrent map access isn't the only problem, there are other critical sections. So I cleaned up the code and used sync.Mutex instead of just protecting the map. Please take a look at the PR and if it's ok I'll complete it with tests.

This particular collector must be concurrent safe just because there's no mechanism protecting the third-party collector on the Prometheus side, as result a couple of concurrent requests to the prometheus handler might crash the app.

In the Prometheus itself the metrics map is protected by the mutex:
https://github.com/prometheus/client_golang/blob/fae2f6306b16a7265d5569c5507623dfaba09053/prometheus/vec.go#L320:323

So any Prometheus collector must be concurrent-safe.

File: exporters/prometheus/exporter.go [hunk 3]
```diff
@@ -59,13 +59,13 @@ var _ metric.Reader = &Exporter{}
...
-	createTargetInfoOnce sync.Once
-	scopeInfos           map[instrumentation.Scope]prometheus.Metric
-	metricFamilies       map[string]*dto.MetricFamily
+	disableTargetInfo bool
+	withoutUnits      bool
+	targetInfo        prometheus.Metric
+	disableScopeInfo  bool
+	scopeInfos        map[instrumentation.Scope]prometheus.Metric
+	metricFamilies    map[string]*dto.MetricFamily
+	mu                sync.RWMutex
```
Comment (Aneurysm9) [inline exporters/prometheus/exporter.go:68]: Can you re-group the members of this struct such that the members protected by the mutex are together underneath the mutex and add a comment indicating as much?

Why a `RWMutex`?  I don't see anywhere that a read lock is taken.

File: exporters/prometheus/exporter.go (hunk 2 above)
Comment (Aneurysm9) [reply exporters/prometheus/exporter.go:125]: Maybe?  In that case it's important to document that `c.validateMetrics()` requires its caller to maintain a lock on the mutex.

**Review by MrAlias — COMMENTED** [review]
This is still missing regression tests.

File: exporters/prometheus/exporter.go (hunk 3 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: Fixed, thanks.

File: exporters/prometheus/exporter.go (hunk 2 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go:126]: Decided to leave it as is because c.reader.Collect (L118) can take some time and it's actually safe.

File: CHANGELOG.md [hunk 4]
```diff
@@ -23,6 +23,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Both the `Histogram` and `HistogramDataPoint` are redefined with a generic argument of `[N int64 | float64]` in `go.opentelemetry.io/otel/sdk/metric/metricdata`. (#3849)
 - The metric `Export` interface from `go.opentelemetry.io/otel/sdk/metric` accepts a `*ResourceMetrics` instead of `ResourceMetrics`. (#3853)
 
+### Fixed
+
+- Panic in Prometheus exporter on concurrent `Collect`. (#3899)
```
Comment (pellared) [inline CHANGELOG.md:28]: Maybe

```suggestion
- Fix a race condition in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
```

1. The `func (c *collector) Collect` is not an publicly visible method 
2. Reference the package instead of name
3. The statement should say what has changed (and not describe the issue)

File: exporters/prometheus/exporter.go [hunk 5]
```diff
@@ -122,16 +123,18 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+		targetInfo, err := createInfoMetric(targetInfoMetricName, targetInfoDescription, metrics.Resource)
 		if err != nil {
 			// If the target info metric is invalid, disable sending it.
 			otel.Handle(err)
 			c.disableTargetInfo = true
 		}
 		c.targetInfo = targetInfo
-	})
+	}
+	c.mu.Unlock()
```
Comment (pellared) [inline exporters/prometheus/exporter.go:136]: This would not be called if `otel.Handle(err)` panics. I think we should do something like:

```go
func () {
  c.mu.Lock()
  defer c.mu.Unlock()
  // critical section
}()
```

File: exporters/prometheus/exporter.go [hunk 6]
```diff
@@ -148,40 +152,48 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 				}
 				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
 			}
+			c.mu.Unlock()
```
Comment (pellared) [inline exporters/prometheus/exporter.go:155]: This would not be called if `otel.Handle(err)` panics. I think we should do something like:

```go
func () {
  c.mu.Lock()
  defer c.mu.Unlock()
  // critical section
}()
```

File: exporters/prometheus/exporter.go (hunk 5 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go:136]: I guess if otel.Handle(err) panics that we don't really care if it would not be called?

File: CHANGELOG.md (hunk 4 above)
Comment (hexdigest) [reply CHANGELOG.md]: Updated, thanks.

File: exporters/prometheus/exporter.go (hunk 6 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: Oh, I see. You mean that if prometheus handler handles handles panic and then we have a deadlock here. I updated the code. PTAL.

File: exporters/prometheus/exporter.go (hunk 5 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: NVM updated.

File: exporters/prometheus/exporter.go (hunk 6 above)
Comment (pellared) [reply exporters/prometheus/exporter.go]: Yup.

The alternative way (probably more bulletproof) would be to move `otel.Handle()` call outside of the critical section. There is a pattern when for concurrent code to "avoid calling unknown/external code while inside a critical section".

This would be to proper fix as it would make having deadlock not possible.

File: exporters/prometheus/exporter.go [hunk 7]
```diff
@@ -140,48 +146,60 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+					if err != nil {
+						otel.Handle(err)
+					}
+					c.scopeInfos[scopeMetrics.Scope] = scopeInfo
 				}
-				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
-			}
-			ch <- scopeInfo
+				ch <- scopeInfo
+			}()
```
Comment (pellared) [inline exporters/prometheus/exporter.go:162]: Sending to a channel inside a critical section looks akward. Is it really needed or can we move `ch <- scopeInfo` outside the critical section? E.g.

```golang
			var prometheus.Metric scopeInfo
			func() {
				c.mu.Lock()
				defer c.mu.Unlock()

				var ( // Declare variables to avoid shadowing.
					ok bool
					err error
				)
				if scopeInfo, ok = c.scopeInfos[scopeMetrics.Scope]; !ok {
					if scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope); err != nil
						otel.Handle(err)
					}
					c.scopeInfos[scopeMetrics.Scope] = scopeInfo
				}
			}()
			ch <- scopeInfo
```

Comment (pellared) [reply exporters/prometheus/exporter.go:162]: But probably we should change it to (based on my other comment)

```golang
			var ( // Declare variables to avoid shadowing.
				prometheus.Metric scopeInfo
				ok bool
				err error
			)

			c.mu.Lock()
			if scopeInfo, ok = c.scopeInfos[scopeMetrics.Scope]; !ok {
				scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope);
				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
			}
			c.mu.Unlock()

			if err != nil {
				otel.Handle(err)
			}
			ch <- scopeInfo
```

Comment (pellared) [reply exporters/prometheus/exporter.go:162]: Maybe it could be further refactored so that the critical section is extracted to a separate method.

Comment (hexdigest) [reply exporters/prometheus/exporter.go:162]: > Maybe it could be further refactored so that the critical section is extracted to a separate method.

I like this one better. Will do.

File: exporters/prometheus/exporter.go [hunk 8]
```diff
@@ -140,48 +146,60 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+					scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope)
+					if err != nil {
+						otel.Handle(err)
+					}
+					c.scopeInfos[scopeMetrics.Scope] = scopeInfo
 				}
-				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
-			}
-			ch <- scopeInfo
+				ch <- scopeInfo
```
[the remark this answers is not in the document]
Comment (hexdigest) [reply exporters/prometheus/exporter.go:161]: I was going to fix this because it seems like a bug to me.

[the remark this answers is not in the document]
Comment (pellared) [reply exporters/prometheus/exporter.go:161]: It would be better to address in a separate PR. I created an issue https://github.com/open-telemetry/opentelemetry-go/issues/3917. Thanks for your help 👍

[the remark this answers is not in the document]
Comment (hexdigest) [reply exporters/prometheus/exporter.go:161]: > It would be better to address in a separate PR. I created an issue #3917. Thanks for your help +1

I have to fix the very same section.

File: exporters/prometheus/exporter.go (hunk 7 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: @pellared refactored, please take a look.

File: exporters/prometheus/exporter.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: @pellared I fixed this too. Kind of understand that it would be better to address this in a separate PR but I'm working on these exact lines of code so it would be weird to leave this as is.

File: exporters/prometheus/exporter.go (hunk 6 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: This is also done.

File: exporters/prometheus/exporter.go [hunk 9]
```diff
@@ -381,16 +375,66 @@ func sanitizeName(n string) string {
...
+	}
+
+	scopeInfo, err := createScopeInfoMetric(scope)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create scope info metric: %w", err)
+	}
+
+	c.mu.Lock()
+	c.scopeInfos[scope] = scopeInfo
+	c.mu.Unlock()
```
Comment (pellared) [inline exporters/prometheus/exporter.go:417]: race condition: this can lead to two `scopeInfo` instances for the same `scope` if two `Collect` are running concurrently.

**Review by pellared — CHANGES_REQUESTED** [review]
I do not feel convinced that we need so many critical sections.

**Comment by hexdigest** [comment]
> I do not feel convinced that we need so many critical sections.

@pellared locking the whole func now. Also I think it was your idea to move otel.Handle out of the critical section.

**Comment by pellared** [comment]
> I think it was your idea to move otel.Handle out of the critical section.

Move `otel.Handle` out (just call it later). Not create multiple small critical sections 😉 See: https://github.com/open-telemetry/opentelemetry-go/pull/3899#discussion_r1144422955

**Comment by hexdigest** [comment]
> > I think it was your idea to move otel.Handle out of the critical section.
> 
> Move `otel.Handle` out (just call it later). Not to create multiple small critical sections wink See: [#3899 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/3899#discussion_r1144422955)

@pellared done this too PTAL

File: CHANGELOG.md [hunk 10]
```diff
@@ -23,6 +23,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Both the `Histogram` and `HistogramDataPoint` are redefined with a generic argument of `[N int64 | float64]` in `go.opentelemetry.io/otel/sdk/metric/metricdata`. (#3849)
 - The metric `Export` interface from `go.opentelemetry.io/otel/sdk/metric` accepts a `*ResourceMetrics` instead of `ResourceMetrics`. (#3853)
 
+### Fixed
+
+- Fix a race condition in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
```
Comment (pellared) [inline CHANGELOG.md:28]: It looks like this PR does more than just that.

Comment (hexdigest) [reply CHANGELOG.md]: Updated, thanks!

File: exporters/prometheus/exporter.go (hunk 9 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go]: Fixed, even though technically it's not a race condition.

File: exporters/prometheus/exporter.go [hunk 11]
```diff
@@ -381,16 +379,64 @@ func sanitizeName(n string) string {
...
+	if ok {
+		return scopeInfo, nil
+	}
+
+	scopeInfo, err := createScopeInfoMetric(scope)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create scope info metric: %w", err)
+	}
+
+	c.scopeInfos[scope] = scopeInfo
```
Comment (pellared) [inline exporters/prometheus/exporter.go:419]: _Not directly related to this PR_

If non-nil error is returned by `createScopeInfoMetric` then it will be called during each `Collect` for given `scope`.

File: exporters/prometheus/exporter.go [hunk 12]
```diff
@@ -381,16 +379,64 @@ func sanitizeName(n string) string {
...
+
+func (c *collector) validateMetrics(name, description string, metricType *dto.MetricType) (drop bool, help string) {
+	c.mu.RLock()
+	emf, exist := c.metricFamilies[name]
+	c.mu.RUnlock()
+
 	if !exist {
-		mfs[name] = &dto.MetricFamily{
+		c.mu.Lock()
+		c.metricFamilies[name] = &dto.MetricFamily{
```
Comment (pellared) [inline exporters/prometheus/exporter.go:431]: The whole method needs to be inside `c.mu.Lock()`. Otherwise they may be two concurrent `Collect` may create two `MetricFamily` and even with `description` and `metricType`.

After this change we can change the `mu` type to regular `sync.Mutex`.

File: exporters/prometheus/exporter.go (hunk 11 above)
Comment (pellared) [reply exporters/prometheus/exporter.go:419]: Do you want to address it in this issue as well or shall I create an issue for it?

**Comment by pellared** [comment]
Side note: I added `Also fixes #3917` to the PR description. I hope you don't mind 😉

File: CHANGELOG.md [hunk 13]
```diff
@@ -23,6 +23,11 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Both the `Histogram` and `HistogramDataPoint` are redefined with a generic argument of `[N int64 | float64]` in `go.opentelemetry.io/otel/sdk/metric/metricdata`. (#3849)
 - The metric `Export` interface from `go.opentelemetry.io/otel/sdk/metric` accepts a `*ResourceMetrics` instead of `ResourceMetrics`. (#3853)
 
+### Fixed
+
+- Fix race conditions in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
+- Fix sending nil `scopeInfo` to metrics channel in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic in `github.com/prometheus/client_golang/prometheus` (#3917)
```
Comment (pellared) [inline CHANGELOG.md:29]: Are you able to add a unit test for this scenario?

Comment (hexdigest) [reply CHANGELOG.md]: @pellared Implemented test to reproduce this issue, in main branch it gives expected test failure (and prometheus panic):

```go
2023/04/28 14:04:16 label value "\xff\xfe\xfd" is not valid UTF-8
2023/04/28 14:04:16 "___" is not a valid label name for metric "__total"
--- FAIL: TestNotNilScopeinfoInCollect (0.00s)
    exporter_test.go:798: 
        	Error Trace:	/tmp/opentelemetry-go/exporters/prometheus/exporter_test.go:798
        	Error:      	Expected value not to be nil.
        	Test:       	TestNotNilScopeinfoInCollect
FAIL
FAIL	go.opentelemetry.io/otel/exporters/prometheus	0.010
```

**Comment by hexdigest** [comment]
Is anybody home? Fixes are here, tests too.

**Review by pellared — CHANGES_REQUESTED** [review]
1. https://github.com/open-telemetry/opentelemetry-go/pull/3899#discussion_r1149002680 is not addressed
2. Lint is failing: https://github.com/open-telemetry/opentelemetry-go/actions/runs/4843560204/jobs/8684967396

File: exporters/prometheus/exporter.go (hunk 12 above)
Comment (hexdigest) [reply exporters/prometheus/exporter.go:436]: Fixed, PTAL

**Comment by hexdigest** [comment]
> 1. [Concurrent collect bugfix #3899 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/3899#discussion_r1149002680) is not addressed
> 2. Lint is failing: https://github.com/open-telemetry/opentelemetry-go/actions/runs/4843560204/jobs/8684967396

@pellared everything is fixed now.

File: exporters/prometheus/exporter.go (hunk 11 above)
Comment (pellared) [reply exporters/prometheus/exporter.go:424]: I created https://github.com/open-telemetry/opentelemetry-go/issues/4091

Feel free to address it in a separate PR 😉

File: CHANGELOG.md [hunk 14]
```diff
@@ -174,6 +174,8 @@ This release drops the compatibility guarantee of [Go 1.18].
 
 - Handle empty environment variable as it they were not set. (#3764)
 - Clarify the `httpconv` and `netconv` packages in `go.opentelemetry.io/otel/semconv/*` provide tracing semantic conventions. (#3823)
+- Fix race conditions in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
+- Fix sending nil `scopeInfo` to metrics channel in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic in `github.com/prometheus/client_golang/prometheus` (#3917)
```
Comment (pellared) [inline CHANGELOG.md:178]:
```suggestion
- Fix sending nil `scopeInfo` to metrics channel in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic in `github.com/prometheus/client_golang/prometheus`. (#3917)
```

**Review by pellared — APPROVED** [review]

**Comment by hexdigest** [comment]
@MrAlias anything else?

**Comment by hexdigest** [comment]
@pellared is there any chance to merge this and fix the panic?

**Comment by pellared** [comment]
> @pellared is there any chance to merge this and fix the panic?

The PR needs to be approved by at least one more non-Splunk approver. Moreover it can be only merged by a maintainer. I will ask others for review.

**Review by dmathieu — APPROVED** [review]

File: CHANGELOG.md [hunk 15]
```diff
@@ -188,6 +188,8 @@ This release drops the compatibility guarantee of [Go 1.18].
 
 - Handle empty environment variable as it they were not set. (#3764)
 - Clarify the `httpconv` and `netconv` packages in `go.opentelemetry.io/otel/semconv/*` provide tracing semantic conventions. (#3823)
+- Fix race conditions in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
+- Fix sending nil `scopeInfo` to metrics channel in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic in `github.com/prometheus/client_golang/prometheus`. (#3917)
```
Comment (pellared) [inline CHANGELOG.md:192]:
```suggestion
- Fix race conditions in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic. (#3899)
- Fix sending nil `scopeInfo` to metrics channel in `go.opentelemetry.io/otel/exporters/metric/prometheus` that could cause a panic in `github.com/prometheus/client_golang/prometheus`. (#3899)
```

**Comment by hexdigest** [comment]
Just curious how many more months it will take to fix several panics in this package.

**Comment by pellared** [comment]
> Just curious how many more months it will take to fix several panics in this package.

@hexdigest I am planning to merge this today as there was enough time for other contributors to review. Can you just please fix `TestNotNilScopeinfoInCollect`?

File: exporters/prometheus/exporter.go [hunk 16]
```diff
@@ -386,16 +387,63 @@ func sanitizeName(n string) string {
...
+
+	switch v := m.Data.(type) {
+	case metricdata.Histogram[int64], metricdata.Histogram[float64]:
+		return dto.MetricType_HISTOGRAM.Enum(), name
+	case metricdata.Sum[float64]:
+		if v.IsMonotonic {
+			return dto.MetricType_COUNTER.Enum(), name + counterSuffix
+		}
+		return dto.MetricType_GAUGE.Enum(), name
+	case metricdata.Sum[int64]:
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:401]: can this case be combined with the other Sum case?

File: exporters/prometheus/exporter.go [hunk 17]
```diff
@@ -124,16 +127,31 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+
+		if c.targetInfo == nil && !c.disableTargetInfo {
+			targetInfo, err = createInfoMetric(targetInfoMetricName, targetInfoDescription, metrics.Resource)
+			if err != nil {
+				// If the target info metric is invalid, disable sending it.
+				c.disableTargetInfo = true
+				return
+			}
+
+			c.targetInfo = targetInfo
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:151]:
```suggestion
	func() {
		c.mu.Lock()
		defer c.mu.Unlock()

		if c.targetInfo == nil && !c.disableTargetInfo {
			targetInfo, err := createInfoMetric(targetInfoMetricName, targetInfoDescription, metrics.Resource)
			if err != nil {
				// If the target info metric is invalid, disable sending it.
				c.disableTargetInfo = true
				otel.Handle(err)
				return
			}

			c.targetInfo = targetInfo
```
Simplify this block.

**Review by dashpole — APPROVED** [review]
Sorry I wasn't able to get to this sooner, i'm just getting back from leave.  This is a great improvement overall, and a needed fix.  Comments are non-blocking.

**Comment by pellared** [comment]
@hexdigest Are you able to fix the build and consider addressing @dashpole comments?

**Comment by hexdigest** [comment]
> @hexdigest Are you able to fix the build and consider addressing @dashpole comments?

Yep, I will

**Comment by hexdigest** [comment]
> @hexdigest Are you able to fix the build and consider addressing @dashpole comments?

The problem is that this is the only trick path I found to reproduce the issue in the test:
```invalid instrument name: ���: must start with a letter```

And now it's being checked somewhere else I suppose, so I might just remove this test.

**Comment by pellared** [comment]
I am fine with removing the test as it is a very edge case and personally I do not know how to reproduce it.

**Comment by pellared** [comment]
@hexdigest I quickly checked https://github.com/prometheus/client_golang/ and tried to fix the test in browser. https://github.com/open-telemetry/opentelemetry-go/pull/3899/commits/fcfb7755faa6ce81a607824e5c12aa18b37a8303

I hope it helps...

File: exporters/prometheus/exporter_test.go [hunk 18]
```diff
@@ -672,3 +676,164 @@ func TestDuplicateMetrics(t *testing.T) {
...
+	ch := make(chan prometheus.Metric)
+
+	go collector.Collect(ch)
+
+	for {
+		select {
+		case m := <-ch:
+			require.NotNil(t, m)
+
+			if strings.Contains(m.Desc().String(), validName) {
```
Comment (pellared) [inline exporters/prometheus/exporter_test.go:832]: Shouldn't we make sure that the "invalid" metric is not emitted?

Also based on test coverage I am not sure if I updated the test correctly.

File: exporters/prometheus/exporter_test.go [hunk 19]
```diff
@@ -672,3 +676,164 @@ func TestDuplicateMetrics(t *testing.T) {
...
+		otelmetric.WithUnit("By"),
+		otelmetric.WithDescription("meter counter foo"))
+	assert.NoError(t, err)
+	fooB.Add(ctx, 100, opt)
+
+	concurrencyLevel := 100
+	ch := make(chan prometheus.Metric, concurrencyLevel)
+
+	for i := 0; i < concurrencyLevel; i++ {
+		go func() {
```
Comment (pellared) [inline exporters/prometheus/exporter_test.go:742]: We should use a `var wg sync.WaitGroup` to make sure that all goroutines have finished.

File: exporters/prometheus/exporter_test.go [hunk 20]
```diff
@@ -672,3 +676,164 @@ func TestDuplicateMetrics(t *testing.T) {
...
+		otelmetric.WithDescription(validName))
+	assert.NoError(t, err)
+
+	counterB.Add(ctx, 100, otelmetric.WithAttributes(
+		attribute.Key(validName).String(validName),
+	))
+
+	ch := make(chan prometheus.Metric)
+
+	go collector.Collect(ch)
```
Comment (pellared) [inline exporters/prometheus/exporter_test.go:825]: Maybe we can use a buffered channel and then we would not need to start a new goroutine.

File: exporters/prometheus/exporter_test.go [hunk 21]
```diff
@@ -672,3 +676,164 @@ func TestDuplicateMetrics(t *testing.T) {
 		})
 	}
 }
+
+func TestConcurrentCollect(t *testing.T) {
```
Comment (pellared) [inline exporters/prometheus/exporter_test.go:680]:
```suggestion
func TestCollectConcurrentSafe(t *testing.T) {
```

(this pattern is widely used)

**Comment by pellared** [comment]
@hexdigest Let us know if you do not have time or do not want to address the comment. I can merge this PR and address the comments myself by creating a new PR.

**Comment by hexdigest** [comment]
Let's merge it now

File: exporters/prometheus/exporter_test.go (hunk 20 above)
Comment (pellared) [reply exporters/prometheus/exporter_test.go:825]: https://github.com/open-telemetry/opentelemetry-go/issues/4270

File: exporters/prometheus/exporter_test.go (hunk 19 above)
Comment (pellared) [reply exporters/prometheus/exporter_test.go:742]: https://github.com/open-telemetry/opentelemetry-go/issues/4270

File: exporters/prometheus/exporter_test.go (hunk 18 above)
Comment (pellared) [reply exporters/prometheus/exporter_test.go:832]: https://github.com/open-telemetry/opentelemetry-go/issues/4270

File: exporters/prometheus/exporter.go (hunk 16 above)
Comment (pellared) [reply exporters/prometheus/exporter.go:394]: https://github.com/open-telemetry/opentelemetry-go/issues/4270

Comment (pellared) [reply exporters/prometheus/exporter.go:394]: @dashpole Do you mean e.g.:

```go
case metricdata.Histogram[int64], metricdata.Histogram[float64]:
				addHistogramMetric(ch, v, m, keys, values, name)
```

?

If so then the compiler cannot handle this case: 

> type metricdata.Aggregation of v does not match metricdata.Histogram[N] (cannot infer N)compiler[CannotInferTypeArgs](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#CannotInferTypeArgs)

Comment (dashpole) [reply exporters/prometheus/exporter.go:394]: Thanks
